### PR TITLE
Fix the locus of foreign upper path

### PR DESCRIPTION
**Pull request by zhaorui20005**
This PR is separated from https://github.com/greenplum-db/gpdb/pull/15525.

When `mpp_execute` is set to 'all segments' and `num_segments` is set to a different number of the default segment number, then we need to generate a path with the correct number of locus and a correct distribution policy for foreign table.
But the legacy code fails to consider the FDW option 'num_segments', so this PR is to fix this issue.

Things this PR will do:
- Add `num_segments` option for postgres_fdw.
- Fix locus of foreign_upper_path.
- Add testing cases.

### CODE REVIEW

**Comment by kainwen** [comment]
Any possible to add a test?

File: src/backend/optimizer/util/pathnode.c [hunk 1]
```diff
@@ -3526,7 +3526,11 @@ create_foreign_upper_path(PlannerInfo *root, RelOptInfo *rel,
 			CdbPathLocus_MakeGeneral(&(pathnode->path.locus));
 			break;
 		case FTEXECLOCATION_ALL_SEGMENTS:
-			CdbPathLocus_MakeStrewn(&(pathnode->path.locus), getgpsegmentCount());
```
Comment (zxuejing) [inline src/backend/optimizer/util/pathnode.c:3529]: can we use pathnode->path.locus.numsegments instead of getgpsegmentCount()?

Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c:3529]: Yeah, actually it is better to use `cdbpathlocus_from_baserel()` if the policy is set correctly for foreign table.

**Comment by zhaorui20005** [comment]
> Any possible to add a test?

Testing case added.

File: src/backend/cdb/cdbcat.c [hunk 2]
```diff
@@ -394,7 +395,10 @@ GpPolicyFetch(Oid tbloid)
 				 * to foreign tables from all segments when the mpp_execute
 				 * option is set to 'all segments'
 				 */
-				return createRandomPartitionedPolicy(getgpsegmentCount());
+				if (server)
+					return createRandomPartitionedPolicy (server->num_segments);
```
Comment (adam8157) [inline src/backend/cdb/cdbcat.c:399]: redundant space character

File: src/backend/optimizer/util/pathnode.c [hunk 3]
```diff
@@ -3375,36 +3375,7 @@ create_foreignscan_path(PlannerInfo *root, RelOptInfo *rel,
...
-			elog(ERROR, "unrecognized exec_location '%c'", rel->exec_location);
-		}
-	}
-	else
-	{
-		/* make entry locus for utility role */
-		CdbPathLocus_MakeEntry(&(pathnode->path.locus));
-	}
-
+	pathnode->path.locus = cdbpathlocus_from_baserel(root, rel);
```
Comment (adam8157) [inline src/backend/optimizer/util/pathnode.c:3378]: Lost the FTEXECLOCATION_ANY here.

Wrap these lines as a function to DRY instead.

File: src/backend/cdb/cdbcat.c (hunk 2 above)
Comment (zhaorui20005) [reply src/backend/cdb/cdbcat.c:399]: Thanks, I will fix this.

File: src/backend/optimizer/util/pathnode.c (hunk 3 above)
Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c]: Correct, the foreign table with mpp_execute set to 'any' will generate different plans compared to 'coordinator'. I have add some testing cases for that.

File: src/backend/optimizer/util/pathnode.c [hunk 4]
```diff
@@ -3375,36 +3391,7 @@ create_foreignscan_path(PlannerInfo *root, RelOptInfo *rel,
...
-			break;
-		case FTEXECLOCATION_COORDINATOR:
-			CdbPathLocus_MakeEntry(&(pathnode->path.locus));
-			break;
-		default:
-			elog(ERROR, "unrecognized exec_location '%c'", rel->exec_location);
-		}
-	}
-	else
-	{
```
Comment (zxuejing) [inline src/backend/optimizer/util/pathnode.c:3403]: if (Gp_role != GP_ROLE_DISPATCH) && (rel->exec_location == FTEXECLOCATION_ANY)
here we make CdbPathLocus_MakeEntry
But after we use get_cdbpathlocus_for_foreign_relation
we will make CdbPathLocus_MakeGeneral

so here we can not delete else branch or add Gp_role judgement in get_cdbpathlocus_for_foreign_relation

Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c:3403]: Sure, but I wonder why the GP_role check does not happen in `create_foreign_join_path()` and `create_foreign_upper_path`, do you know the reason?

Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c:3403]: Actually cdbpathlocus_from_baserel will still make CdbPathLocus_MakeEntry locus for it.

File: src/backend/optimizer/util/pathnode.c [hunk 5]
```diff
@@ -3338,6 +3338,22 @@ path_contains_inner_index(Path *path)
 	return false;
 }
 
+/* Set locus for foreign path node */
+static void
+get_cdbpathlocus_for_foreign_relation(struct PlannerInfo   *root,
```
Comment (adam8157) [inline src/backend/optimizer/util/pathnode.c:3343]: nitpick, `make_cdbpathlocus_for_foreign_relations`

File: src/backend/optimizer/util/pathnode.c [hunk 6]
```diff
@@ -3338,6 +3338,22 @@ path_contains_inner_index(Path *path)
...
+	if (rel->exec_location == FTEXECLOCATION_ANY)
+	{
+		CdbPathLocus_MakeGeneral(&(pathnode->path.locus));
+	}
+	else
+	{
+		pathnode->path.locus = cdbpathlocus_from_baserel(root, rel);
+	}
+	return;
+}
```
Comment (adam8157) [inline src/backend/optimizer/util/pathnode.c:3362]: nitpick, add a blank line here.

File: src/backend/optimizer/util/pathnode.c (hunk 4 above)
Comment (adam8157) [reply src/backend/optimizer/util/pathnode.c:3403]: This case for utility mode was introduced by the GP2GP feature, the change is likely fine, please have it tested (manually is OK).

**Review by adam8157 — APPROVED** [review]
Except for the tiny issues, LGTM

File: src/backend/optimizer/util/pathnode.c (hunk 5 above)
Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c:3343]: Thanks, `make_cdbpathlocus_for_foreign_relations` is better.

File: src/backend/optimizer/util/pathnode.c (hunk 6 above)
Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c:3362]: OK, I will fix this.

File: src/backend/optimizer/util/pathnode.c (hunk 4 above)
Comment (zhaorui20005) [reply src/backend/optimizer/util/pathnode.c:3403]: Yes, I have tested it manually with utility mode, in this final version, all locus is set to entry in utility mode.

File: contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out [hunk 7]
```diff
@@ -1223,3 +1279,71 @@ select * from part_mixed_dpe, non_part where part_mixed_dpe.b=non_part.b;
...
+   ->  Result
+         ->  Redistribute Motion 3:3  (slice1; segments: 3)
+               Hash Key: t1.a
+               ->  Hash Join
+                     Hash Cond: (t1.b = a)
+                     ->  Redistribute Motion 3:3  (slice2; segments: 3)
+                           Hash Key: t1.b
+                           ->  Seq Scan on t1
+                     ->  Hash
+                           ->  GroupAggregate
```
Comment (avamingli) [inline contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out:1309]: Just curious, why ORCA doesn't use Hash Semi Join for the two cases to handle `IN` , like the first case?

Comment (zhaorui20005) [reply contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out:1309]: Hi, Mingli, long time no see. I have no idea why is the plan looks like this. I just want to make a query that differs according to the locus of foreign tables.

**Review by avamingli — APPROVED** [review]

**Pull request merged**